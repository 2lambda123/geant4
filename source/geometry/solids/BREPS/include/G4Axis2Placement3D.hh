// This code implementation is the intellectual property of
// the GEANT4 collaboration.
//
// By copying, distributing or modifying the Program (or any work
// based on the Program) you indicate your acceptance of this statement,
// and all its terms.
//
// $Id: G4Axis2Placement3D.hh,v 1.3 2000-11-08 14:21:58 gcosmo Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//
// ----------------------------------------------------------------------
// Class G4Axis2Placement3D
//
// Class description:
//
// Class defining an axis placed in a 3D space.
// Attributes are: a point location, an axis vector with its
// reference direction and coordinates in space.

// Authors: J.Sulkimo, P.Urban.
// Revisions by: L.Broglia, G.Cosmo.
// ----------------------------------------------------------------------
#ifndef __G4Placement3D_h
#define __G4Placement3D_h 1

#include "G4Point3D.hh"
#include "G4Vector3D.hh"
#include "G4Transform3D.hh"
#include "G4PointRat.hh"
#include "G4Ray.hh"

class G4Axis2Placement3D 
{

public: // with description

  G4Axis2Placement3D();
  G4Axis2Placement3D( const G4Vector3D& refDirection0 ,
		      const G4Vector3D& axis0         , 
		      const G4Point3D&  location0      );
  ~G4Axis2Placement3D();
    // Constructors & destructor
  
  G4Axis2Placement3D(const G4Axis2Placement3D&);
  G4Axis2Placement3D& operator=(const G4Axis2Placement3D&);
    // Copy constructor and assignment operator.

  inline G4bool operator==(const G4Axis2Placement3D& other) const;
    // Equality operator.

  inline void Init( const G4Vector3D& refDirection0 , 
	            const G4Vector3D& axis0         ,
	            const G4Point3D&  location0      );

  inline G4Point3D  GetLocation() const;
  inline G4Vector3D GetAxis() const;
  inline G4Vector3D GetRefDirection() const;
    // Get/Set for geometric data

  inline G4Vector3D GetPX() const;
  inline G4Vector3D GetPY() const;
  inline G4Vector3D GetPZ() const;
    // Placement coordinate axes

  inline const G4Transform3D& GetToPlacementCoordinates() const;
  inline const G4Transform3D& GetFromPlacementCoordinates() const; 
    // Transformation from/to the placement coordinate system

public:

  //inline void Project (G4ThreeVec& Coord, const G4ThreeVec& Pt2, 
  //                     const G4Plane& Pl1, const G4Plane& Pl2)
  // {
  //   Coord.X(Pt2.X()*Pl1.a + Pt2.Y()*Pl1.b + Pt2.Z()*Pl1.c - Pl1.d);
  //   Coord.Y(Pt2.X()*Pl2.a + Pt2.Y()*Pl2.b + Pt2.Z()*Pl2.c - Pl2.d);
  //   Coord.Z(0);
  // }

private:

  // geometric data
  G4Point3D location;
  G4Vector3D axis;
  G4Vector3D refDirection;  

  // placement coordinate axes
  G4Vector3D pX, pY, pZ;

  G4Transform3D toPlacementCoordinates;
  G4Transform3D fromPlacementCoordinates;

};

#include "G4Axis2Placement3D.icc"

#endif
