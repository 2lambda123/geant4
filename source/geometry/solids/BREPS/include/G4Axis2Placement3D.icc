// This code implementation is the intellectual property of
// the GEANT4 collaboration.
//
// By copying, distributing or modifying the Program (or any work
// based on the Program) you indicate your acceptance of this statement,
// and all its terms.
//
// $Id: G4Axis2Placement3D.icc,v 1.2 2000-08-28 08:57:41 gcosmo Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//
// --------------------------------------------------------------------
// GEANT 4 inline definitions file
//
// G4AxisPlacement3D.icc
//
// Implementation of inline methods of G4AxisPlacement3D
// --------------------------------------------------------------------

inline void
G4Axis2Placement3D::Init( const G4Vector3D& refDirection0 ,
			  const G4Vector3D& axis0         ,
			  const G4Point3D& location0       ) 
{
  refDirection = refDirection0;
  axis         = axis0;
  location     = location0;
  
  // get the axes of the placement coordinate system
  // (p[] of the STEP standard)
  pZ = axis.unit();
  pX = (refDirection-(refDirection*pZ)*pZ).unit();
  pY = pZ.cross(pX); // normalized
  
  // basis transformation
  fromPlacementCoordinates= HepTranslate3D(location)
    * G4Transform3D(HepXHat, HepYHat, HepZHat,
		     pX, pY, pZ);

  toPlacementCoordinates= fromPlacementCoordinates.inverse();
}


inline
G4Axis2Placement3D::G4Axis2Placement3D(const G4Vector3D& refDirection0,
				       const G4Vector3D& axis0        ,
				       const G4Point3D& location0      )
{
  Init( refDirection0, axis0, location0);
}


inline G4Point3D
G4Axis2Placement3D::GetLocation() const
{
  return location;
}

inline G4Vector3D
G4Axis2Placement3D::GetAxis() const
{
  return axis;
}

inline G4Vector3D
G4Axis2Placement3D::GetRefDirection() const 
{
  return refDirection;
}

/////////////////////////////////////////////////////////////////////////////

inline const G4Transform3D&
G4Axis2Placement3D::GetToPlacementCoordinates() const 
{
  return toPlacementCoordinates;
}

inline const G4Transform3D&
G4Axis2Placement3D::GetFromPlacementCoordinates() const 
{
  return fromPlacementCoordinates;
}

inline G4Vector3D
G4Axis2Placement3D::GetPX() const
{
  return pX;
}

inline G4Vector3D
G4Axis2Placement3D::GetPY() const
{
  return pY;
}

inline G4Vector3D
G4Axis2Placement3D::GetPZ() const
{
  return pZ;
}
