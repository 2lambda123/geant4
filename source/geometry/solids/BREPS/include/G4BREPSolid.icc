// This code implementation is the intellectual property of
// the GEANT4 collaboration.
//
// By copying, distributing or modifying the Program (or any work
// based on the Program) you indicate your acceptance of this statement,
// and all its terms.
//
// $Id: G4BREPSolid.icc,v 1.3 2000-11-08 14:21:58 gcosmo Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//
// --------------------------------------------------------------------
// GEANT 4 inline definitions file
//
// G4BREPSolid.icc
//
// Implementation of inline methods of G4BREPSolid
// --------------------------------------------------------------------

inline G4double G4BREPSolid::IntersectionDistance() const
{
  return intersectionDistance;
}

inline void G4BREPSolid::IntersectionDistance(G4double d) const 
{
  ((G4BREPSolid*)this)->intersectionDistance=d;
}   

inline G4Surface* G4BREPSolid::GetSurface(G4int nr) const
{
  return SurfaceVec[nr];
}
  
inline void G4BREPSolid::Active(G4int x) const 
{
  ((G4BREPSolid*)this)->active=x;
}
  
inline G4int G4BREPSolid::Active() const
{
  return active;
}

inline G4double G4BREPSolid::GetShortestDistance() const
{
  return ShortestDistance;
}

inline G4int G4BREPSolid::GetId() const
{
  return Id;
}

inline void G4BREPSolid::SetId(G4int n)
{
  Id = n;
}

inline const G4String& G4BREPSolid::GetName() const
{
  return solidname;
}

inline void G4BREPSolid::SetName(const G4String& name)
{
  solidname = name;
}

inline G4int G4BREPSolid::GetNumberOfSolids() const
{
  return NumberOfSolids;
}

inline G4int G4BREPSolid::GetNumberOfFaces() const
{
  return nb_of_surfaces;
}

inline const G4Axis2Placement3D* G4BREPSolid::GetPlace() const
{
  return place;
}

inline const G4BoundingBox3D* G4BREPSolid::GetBBox() const
{
  return bbox;
} 

inline G4int G4BREPSolid::StartInside() const 
{
  return startInside;
}
  
inline void G4BREPSolid::StartInside(G4int si) const 
{
  ((G4BREPSolid*)this)->startInside=si;
}  

inline void G4BREPSolid::QuickSort( register G4Surface** SrfVec, 
		                    register G4int left,
				    register G4int right) const
{
    register G4int i=left;
    register G4int j=right;
    register G4Surface* elem1;
    register G4Surface* elem2 = SrfVec[(left+right)/2];
    register G4double tmpdistance;
    do
    {
      tmpdistance = elem2->GetDistance();
      while ( SrfVec[i]->GetDistance() < tmpdistance  && i < right ) i++;
      while (tmpdistance < SrfVec[j]->GetDistance()  && j > left ) j--;
      if(i<=j)
      {
	elem1 = SrfVec[i];
	SrfVec[i] = SrfVec[j];
	SrfVec[j] = elem1;
	i++;j--;
      }
    } while (i<=j);
      
    if( left < j  ) QuickSort(SrfVec,left, j );
    if( i < right ) QuickSort(SrfVec,i, right);    
}
