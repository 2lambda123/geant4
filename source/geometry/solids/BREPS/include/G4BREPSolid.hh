#ifndef __SOLID_H
#define __SOLID_H
#include "G4VSolid.hh"	      
#include "G4VisExtent.hh"      
#include "G4Surface.hh"
#include "G4Axis2Placement3D.hh"
#include "G4PointRat.hh"	 
#include "G4BoundingBox3D.hh"	 

class STEPentity;
class InstMgr;
class G4Ray;


class G4BREPSolid : public G4VSolid
{

public:

  G4BREPSolid(const G4String name);
  G4BREPSolid(const G4String, G4Surface**, G4int);
  ~G4BREPSolid();

  virtual G4String GetEntityType() const {return "Closed_Shell";}  
  virtual void Initialize();
  G4int CreateSTEPData(); // not yet implemented 

  G4bool CalculateExtent(const EAxis              pAxis      ,
			 const G4VoxelLimits&     pVoxelLimit,
			 const G4AffineTransform& pTransform ,
			 G4double&                pMin       , 
			 G4double&                pMax        ) const;

  virtual EInside Inside(register const G4ThreeVector&) const;

  virtual G4ThreeVector SurfaceNormal(const G4ThreeVector&) const;

  virtual G4double DistanceToIn(const G4ThreeVector&) const;
  virtual G4double DistanceToIn(register const G4ThreeVector&,
				register const G4ThreeVector&) const;

  virtual G4double DistanceToOut(const G4ThreeVector&) const;
  virtual G4double DistanceToOut(register const G4ThreeVector&,
				 register const G4ThreeVector&,
				 const G4bool  calcNorm=false , 
				 G4bool        *validNorm=0   , 
				 G4ThreeVector *n=0             ) const;
 

  G4Point3D Scope(); // ???

  void DescribeYourselfTo (G4VGraphicsScene& scene) const;
  G4VisExtent   GetExtent        () const;
  G4Polyhedron* CreatePolyhedron () const;
  G4NURBS*      CreateNURBS      () const;

  G4int Intersect(register const G4Ray&)const;

  inline G4double IntersectionDistance()const{return intersectionDistance;}
  void IntersectionDistance(const G4double d)const 
  {
    ((G4BREPSolid*)this)->intersectionDistance=d;
  }   

  G4Surface* GetSurface(G4int nr)
  {
    return SurfaceVec[nr];
  }
  
  inline void Active(const G4int x)const 
  {
    ((G4BREPSolid*)this)->active=x;
  }
  
  inline G4int Active() const {return active;}
  
  virtual inline void Reset() const
  {
    ((G4BREPSolid*)this)->active=1;
    ((G4BREPSolid*)this)->intersectionDistance=kInfinity;
    ((G4BREPSolid*)this)->startInside=0; 
     
    for(register G4int a=0;a<nb_of_surfaces;a++)
      SurfaceVec[a]->Reset();

    ShortestDistance = kInfinity;
  }

  static G4int NumberOfSolids;
  static InstMgr InstanceList;

  G4double GetShortestDistance() const {return ShortestDistance;}

  G4int GetId() const {return Id;}

  void SetId(G4int id) {Id = id;}

  G4String GetName() const {return solidname;}

  void SetName(G4String name) {solidname = name;}

  G4int NumberOfFaces() const {return nb_of_surfaces;} 

  // Add by L. Broglia
  G4Axis2Placement3D* GetPlace() { return place; }
  G4BoundingBox3D*    GetBBox()  { return bbox;  } 
 
protected:

  G4bool  IsConvex();
  virtual void CalcBBoxes();
  void    CheckSurfaceNormals();
  void    RemoveHiddenFaces(register const G4Ray& G4Rayref, G4int)const;   
  void    TestSurfaceBBoxes(register const G4Ray&) const;

  inline G4int StartInside() const 
  {
    return startInside;
  }
  
  inline void StartInside(const G4int si) const 
  {
    ((G4BREPSolid*)this)->startInside=si;
  }  


private:

  G4int IsBox();
  G4int FinalEvaluation(register const G4Ray&, const G4int =0) const;


protected:
  G4Axis2Placement3D* place;
  static G4Ray        Track;
  static G4double     ShortestDistance;
  G4int               Box, Convex, AxisBox, PlaneSolid;
  G4BoundingBox3D*    bbox;   
  G4double            intersectionDistance;
  G4int               active;
  G4int               startInside;
  G4int               nb_of_surfaces;
  G4Point3D           intersection_point;
  G4Surface**         SurfaceVec;
  G4double            RealDist;
  G4String            solidname; 
  G4int               Id;
   

  void QuickSort( register G4Surface** SrfVec, 
		  register G4int left, register G4int right) const
  {
    register G4int i=left;
    register G4int j=right;
    register G4Surface* elem1;
    register G4Surface* elem2 = SrfVec[(left+right)/2];
    register G4double tmpdistance;
    do
    {
      tmpdistance = elem2->Distance();
      while ( SrfVec[i]->Distance() < tmpdistance  && i < right ) i++;
      while (tmpdistance < SrfVec[j]->Distance()  && j > left ) j--;
      if(i<=j)
      {
	elem1 = SrfVec[i];
	SrfVec[i] = SrfVec[j];
	SrfVec[j] = elem1;
	i++;j--;
      }
     } while (i<=j);
      
     if( left < j  ) QuickSort(SrfVec,left, j );
     if( i < right ) QuickSort(SrfVec,i, right);    
    }

};

#endif









