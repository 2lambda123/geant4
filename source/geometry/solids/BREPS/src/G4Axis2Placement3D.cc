// This code implementation is the intellectual property of
// the GEANT4 collaboration.
//
// By copying, distributing or modifying the Program (or any work
// based on the Program) you indicate your acceptance of this statement,
// and all its terms.
//
// $Id: G4Axis2Placement3D.cc,v 1.4 2000-11-10 17:41:29 gcosmo Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//
// ----------------------------------------------------------------------
// GEANT 4 class source file
//
// G4Axis2Placement3D.cc
//
// ----------------------------------------------------------------------

#include "G4Axis2Placement3D.hh"

//G4Axis2Placement3D
G4Axis2Placement3D::G4Axis2Placement3D(){}
G4Axis2Placement3D::~G4Axis2Placement3D(){}

// copy constructor (used in STEPinterface module)
//
G4Axis2Placement3D::G4Axis2Placement3D(const G4Axis2Placement3D& place)
  : refDirection(place.refDirection), 
    axis(place.axis), location(place.location),
    pX(place.pX), pY(place.pY), pZ(place.pZ),
    toPlacementCoordinates(place.toPlacementCoordinates),
    fromPlacementCoordinates(place.fromPlacementCoordinates)
{
}

// assignment operator
//
G4Axis2Placement3D&
G4Axis2Placement3D::operator=(const G4Axis2Placement3D& place)
{
  if (&place == this) return *this;
  
  refDirection             = place.refDirection; 
  axis                     = place.axis;
  location                 = place.location;
  pX                       = place.pX;
  pY                       = place.pY;
  pZ                       = place.pZ;
  toPlacementCoordinates   = place.toPlacementCoordinates;
  fromPlacementCoordinates = place.fromPlacementCoordinates; 

  return *this;
}

/* everything below here is commented-out ...

G4Axis2Placement3D::G4Axis2Placement3D(const G4ThreeVec Dir, 
                                       const G4ThreeVec Axis, 
				       const G4Point3d Pt    )
{
  dir=Dir;
  axis=Axis;
  srf_point=Pt;
  ComputeNormal();
  G4Point3d Pt2 = Pt+Dir;
  G4Point3d Pt3 = Pt+Axis;  
  G4Ray::CalcPlane3Pts(Pl, Pt, Pt2, Pt3);  
}

G4Axis2Placement3D::G4Axis2Placement3D(const G4ThreeVec Dir,
                                       const G4ThreeVec Axis,
				       const G4Point3d Pt1,
				       const G4Point3d Pt2,
				       const G4Point3d Pt3)
{
  dir=Dir;
  axis=Axis;
  srf_point=Pt1;
  ComputeNormal();
  G4Ray::CalcPlane3Pts(Pl, Pt1, Pt2, Pt3);  
}

void
G4Axis2Placement3D::ProjectPlacement(const G4Plane& Pl1,
                                     const G4Plane& Pl2)
{
  Project(ProjectedDir, dir, Pl1, Pl2);
  Project(ProjectedAxis, axis, Pl1, Pl2);
  Project(ProjectedSrfPoint, srf_point, Pl1, Pl2);
  Project(ProjectedNormal, Normal, Pl1, Pl2);
}

void
G4Axis2Placement3D::ComputeNormal()
{

  if(dir == axis)
    Normal = dir;
  else
    {
      Normal.X(dir.Y()*axis.Z() - dir.Z()*axis.Y());
      Normal.Y(dir.X()*axis.Z()- dir.Z()*axis.X());
      Normal.Z(dir.X()*axis.Y() - dir.Y()*axis.X());
    }
}


G4Point3d
G4Axis2Placement3D::EvaluateIntersection(register const G4Ray& rray)
{

// s is solution, line is p + tq, n is G4Plane Normal, r is point on G4Plane 
// all parameters are pointers to arrays of three elements 

    register G4double a, b, t;
    register const G4ThreeVec& RayDir   = rray.GetDir();
    register const G4Point3d& RayStart = rray.GetStart();
    G4double dirx =  RayDir.X();
    G4double diry =  RayDir.Y();
    G4double dirz =  RayDir.Z();
    b = Normal.X() * dirx + Normal.Y() * diry + Normal.Z() * dirz;

    if (fabs(b) < 0.001)//== 0.0)
       // or some better test involving a small positive e     
    {
//    G4cout << "\nLine is parallel to G4Plane.No Hit.";
      G4Point3d hit_point( kInfinity, kInfinity, kInfinity);
      return hit_point;
    }
    G4double startx =  RayStart.X();
    G4double starty =  RayStart.Y();
    G4double startz =  RayStart.Z();    
    
    a = Normal.X() * (srf_point.X() - startx)
      + Normal.Y() * (srf_point.Y() - starty)
      + Normal.Z() * (srf_point.Z() - startz);

    t = a/b;

    // substitute t into line equation
    // to calculate final solution     
    G4Point3d hit_point(startx + t * dirx,starty
                               + t * diry,startz
			       + t * dirz);    
    
//  G4cout << "\nPLANE HIT POINT :" << hit_point.X()
//         << "  " << hit_point.Y() << "  " << hit_point.Z();
    return hit_point;
}
*/
