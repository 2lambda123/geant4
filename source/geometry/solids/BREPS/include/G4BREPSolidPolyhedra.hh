//
// ********************************************************************
// * DISCLAIMER                                                       *
// *                                                                  *
// * The following disclaimer summarizes all the specific disclaimers *
// * of contributors to this software. The specific disclaimers,which *
// * govern, are listed with their locations in:                      *
// *   http://cern.ch/geant4/license                                  *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.                                                             *
// *                                                                  *
// * This  code  implementation is the  intellectual property  of the *
// * GEANT4 collaboration.                                            *
// * By copying,  distributing  or modifying the Program (or any work *
// * based  on  the Program)  you indicate  your  acceptance of  this *
// * statement, and all its terms.                                    *
// ********************************************************************
//
//
// $Id: G4BREPSolidPolyhedra.hh,v 1.10 2002-11-06 23:28:50 radoone Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//
// ----------------------------------------------------------------------
// Class G4BREPSolidPolyhedra
//
//  Class description:
// 
//  The polygonal solid G4BREPSolidPolyhedra is a shape defined by an inner 
//  and outer polygonal surface and two planes perpendicular to the Z axis. 
//  Each polygonal surface is created by linking a series of polygons created 
//  at different planes perpendicular to the Z-axis. All these polygons all 
//  have the same number of sides (sides) and are defined at the same Z planes 
//  for both inner and outer polygonal surfaces. 
//
//  G4BREPSolidPolyhedra( const G4String& name,
//                              G4double  phi1,
//                              G4double  dphi,
//                              G4int     sides,
//                              G4int     num_z_planes,      
//                              G4double  z_start,
//                              G4double  z_values[],
//                              G4double  RMIN[],
//                              G4double  RMAX[] )

// Authors: J.Sulkimo, P.Urban.
// Revisions by: L.Broglia, S.Giani, G.Cosmo.
// ----------------------------------------------------------------------
#ifndef __G4BREPPOLYHEDRA
#define __G4BREPPOLYHEDRA

#include "G4BREPSolid.hh"

class G4BREPSolidPolyhedra : public G4BREPSolid
{
public:  // with description

  G4BREPSolidPolyhedra( const G4String& name,
			G4double phi1,
			G4double dphi,
			G4int    sides,
			G4int    num_z_planes,      
			G4double z_start,
			G4double z_values[],
			G4double RMIN[],
			G4double RMAX[] );
    // Constructor defining the polyhedra through its
    // polygonal surfaces and planes.

  ~G4BREPSolidPolyhedra();
    // Destructor.

  void Initialize();
    // Calculates the bounding box.

  EInside Inside(register const G4ThreeVector& Pt) const;
    // Determines if the point Pt is inside, outside or on the surface
    // of the solid.

  G4ThreeVector SurfaceNormal(const G4ThreeVector&) const;
    // Calculates the normal of the surface at a point on the surface
    // The sense of the normal depends on the sense of the surface.

  G4double DistanceToIn(const G4ThreeVector&) const;
    // Calculates the shortest distance ("safety") from a point
    // outside the solid to any boundary of this solid.
    // Return 0 if the point is already inside.

  G4double DistanceToIn(register const G4ThreeVector& Pt, 
			register const G4ThreeVector& V) const;
    // Calculates the distance from a point Pt outside the solid
    // to the solid's boundary along a specified direction vector V.
    // Note: Intersections with boundaries less than the tolerance must
    //       be ignored if the direction is away from the boundary.

  G4double DistanceToOut(register const G4ThreeVector& Pt, 
			 register const G4ThreeVector& V, 
			 const G4bool calcNorm=false, 
			 G4bool *validNorm=0, G4ThreeVector *n=0) const;
    // Calculates the distance from a point inside the solid to the solid`s
    // boundary along a specified direction vector.
    // Return 0 if the point is already outside.
    // Note: If the shortest distance to a boundary is less than the
    //       tolerance, it is ignored. This allows for a point within a
    //       tolerant boundary to leave immediately.

  G4double DistanceToOut(const G4ThreeVector&) const;
    // Calculates the shortest distance ("safety") from a point inside the
    // solid to any boundary of this solid.
    // Return 0 if the point is already outside.	

  virtual G4std::ostream& StreamInfo(G4std::ostream& os) const;
    // Streams solid contents to output stream.

public:

  G4Polyhedron* CreatePolyhedron () const;
    // Creates a G4Polyhedron

  void Reset() const;
    // Resets all distance attributes.

private:

  G4BREPSolidPolyhedra(const G4BREPSolidPolyhedra&);
  G4BREPSolidPolyhedra& operator=(const G4BREPSolidPolyhedra&);
    // Private copy constructor and assignment operator.
  
  typedef enum _SF {
    EInverse = 0,
    ENormal = 1
  } ESurfaceSense;
    
  G4Surface* CreateTrapezoidalSurface( G4double r1, G4double r2,
                                       const G4Point3D& origin, G4double zDistance,
                                       G4Vector3D& xAxis, G4double partAngle,
                                       ESurfaceSense sense );
  
  G4Surface* CreateTriangularSurface( G4double r1, G4double r2,
                                      const G4Point3D& origin, G4double zDistance,
                                      G4Vector3D& xAxis, G4double partAngle,
                                      ESurfaceSense sense );
  
  G4Surface* ComputePlanarSurface( G4double r1, G4double r2,
                                   const G4Point3D& origin, G4Vector3D& xAxis,
                                   G4int sides, G4double partAngle,
                                   ESurfaceSense sense );

  //   The following is only utilised in storing the shape parameters for
  //  use in visualising this shape.  J.A. Feb  24, 1997
  //
  // R. Chytracek, Nov 2002, Update to new IO dumping mechanism
  
  struct {
    G4double  start_angle;
    G4double  opening_angle;
    G4int     sides;
    G4int     num_z_planes;
    G4double  z_start;               
    G4double* z_values;
    G4double* RMIN;
    G4double* RMAX;
  } constructorParams;
      
};

#endif
