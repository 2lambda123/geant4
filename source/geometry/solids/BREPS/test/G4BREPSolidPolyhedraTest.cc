//////////////////////////////////////////////////////////////////////////
// $Id: G4BREPSolidPolyhedraTest.cc,v 1.6 2000-08-28 08:58:04 gcosmo Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//////////////////////////////////////////////////////////////////////////
//
// This code implementation is the intellectual property of
// the GEANT4 collaboration.
//
// By copying, distributing or modifying the Program (or any work
// based on the Program) you indicate your acceptance of this statement,
// and all its terms.
//
// BREP solid test, create by L. Broglia, 20/10/98
// modification of old G4Gerep test
//



#include "G4Timer.hh"
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include "g4std/fstream"
#include "G4ios.hh" 
#include "G4Axis2Placement3D.hh"
#include "G4BREPSolid.hh"
#include "G4BREPSolidPolyhedra.hh"


G4int main(G4int argc, char **argv)
{
  G4Timer timer;
  
  double RMINVec[5];
  RMINVec[0] = 10;
  RMINVec[1] = 10;
  RMINVec[2] = 60;
  RMINVec[3] = 30;
  RMINVec[4] = 30;

  double RMAXVec[5];
  RMAXVec[0] = 50;
  RMAXVec[1] = 50;
  RMAXVec[2] = 100;
  RMAXVec[3] = 100; 
  RMAXVec[4] = 80;
   
  double Z_Values[5];
  Z_Values[0] = 0;
  Z_Values[1] = 10;  
  Z_Values[2] = 20;
  Z_Values[3] = 30;
  Z_Values[4] = 40;


  G4cout << "\n=======     PolyGon test      ========"<<G4endl;

  G4BREPSolidPolyhedra *MyPGone = new G4BREPSolidPolyhedra ("MyPolyhedra",
							    0            ,
							    2*pi         ,
							    4            ,
							    5            ,
							    0            ,
							    Z_Values     ,
							    RMINVec      ,
							    RMAXVec       );
  G4cout << "\n\nPgon (G4BREPSolid-Polyhedra) created ! "<<G4endl;
  // -> Check methods :
  //  - Inside
  //  - DistanceToIn
  //  - DistanceToOut

  
  EInside in;
  
  G4cout<<"\n\n==================================================";
  G4Point3D  pt(0, -110, 20);
  for (G4int y = -110; y<=110; y+=10)
  {
    pt.setY(y);
    in = MyPGone->Inside(pt);
    
    G4cout << "\nx=" << pt.x() << "  y=" << pt.y() << "  z=" << pt.z();
    
    if( in == kInside )
      G4cout <<" is inside";
    else
      if( in == kOutside )
	G4cout <<" is outside";
      else
	G4cout <<" is on the surface";
  }

  G4cout<<"\n\n==================================================";
  G4Point3D  start( 0, 0, -5);
  G4Vector3D dir1(1, 0, 0);
  G4Vector3D dir2(1, 1, 0);
  G4double   d1, d2;
  G4double x, z;
  
  G4cout<<"\nPdep is (0, 0, z)";
  G4cout<<"\nDir1 is (1, 0, 0)\n";
  G4cout<<"\nDir2 is (1, 1, 0)\n";

  for(z=-5; z<=45; z+=5)
  {
    start.setZ(z);

    in = MyPGone->Inside(start);
    G4cout<< "x=" << start.x() << "  y=" << start.y() << "  z=" << start.z();
    
    if( in == kInside )
    {
      G4cout <<" is inside";

      d1 = MyPGone->DistanceToOut(start, dir1);
      G4cout<<"  distance to out1 ="<<d1;
      d2 = MyPGone->DistanceToOut(start, dir2);
      G4cout<<"  distance to out2 ="<<d2<<G4endl;
    }
    else if( in == kOutside )
    {
      G4cout <<" is outside";

      d1 = MyPGone->DistanceToIn(start, dir1);
      G4cout<<"  distance to in1 ="<<d1;
      d2 = MyPGone->DistanceToIn(start, dir2);
      G4cout<<"  distance to in2 ="<<d2<<G4endl;
    }
    else
      G4cout <<" is on the surface"<< G4endl;
  }
 
  G4cout<<"\n\n==================================================";
  G4Point3D  start3( -110, -110, -5);
  G4Vector3D dir3( 1, 1, 0);
  G4double   d3;
  
  G4cout<<"\nPdep is (-110, -110, z)";
  G4cout<<"\nDir is (1, 1, 0)\n";

  for(z=-5; z<=45; z+=5)
  {
    start3.setZ(z);

    in = MyPGone->Inside(start3);
    G4cout<<"x=" << start3.x() << "  y=" << start3.y() << "  z=" << start3.z();
    
    if( in == kInside )
    {
      G4cout <<" is inside";

      d3 = MyPGone->DistanceToOut(start3, dir3);
      G4cout<<"  distance to out ="<<d3<<G4endl;
    }
    else if( in == kOutside )
    {
      G4cout <<" is outside";

      d3= MyPGone->DistanceToIn(start3, dir3);
      G4cout<<"  distance to in ="<<d3<<G4endl;
    }
    else
      G4cout <<" is on the surface"<< G4endl;
  }

  G4cout << G4endl << G4endl;
}

