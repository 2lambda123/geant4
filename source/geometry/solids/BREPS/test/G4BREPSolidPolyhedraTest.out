
=======     PolyGon test      ========
sf: 34 nb_of_surfaces: 34 Count: 34 zeroCount: 0
Parameters of the G4PGone MyPolyhedra
  starting angle =0
  opening angle =6.28319
  sides =4
  nb of z planes=5
   Z[0] = 0   Rmin[0] = 10   Rmax[0] = 50
   Z[1] = 10   Rmin[1] = 10   Rmax[1] = 50
   Z[2] = 20   Rmin[2] = 60   Rmax[2] = 100
   Z[3] = 30   Rmin[3] = 30   Rmax[3] = 100
   Z[4] = 40   Rmin[4] = 30   Rmax[4] = 80


Pgon (G4BREPSolid-Polyhedra) created ! 


==================================================
x=0  y=-110  z=20 is outside
x=0  y=-100  z=20 is on the surface
x=0  y=-90  z=20 is inside
x=0  y=-80  z=20 is inside
x=0  y=-70  z=20 is inside
x=0  y=-60  z=20 is on the surface
x=0  y=-50  z=20 is outside
x=0  y=-40  z=20 is outside
x=0  y=-30  z=20 is outside
x=0  y=-20  z=20 is outside
x=0  y=-10  z=20 is outside
x=0  y=0  z=20 is outside
x=0  y=10  z=20 is outside
x=0  y=20  z=20 is outside
x=0  y=30  z=20 is outside
x=0  y=40  z=20 is outside
x=0  y=50  z=20 is outside
x=0  y=60  z=20 is on the surface
x=0  y=70  z=20 is inside
x=0  y=80  z=20 is inside
x=0  y=90  z=20 is inside
x=0  y=100  z=20 is on the surface
x=0  y=110  z=20 is outside

==================================================
Pdep is (0, 0, z)
Dir1 is (1, 0, 0)

Dir2 is (1, 1, 0)
x=0  y=0  z=-5 is outside  distance to in1 =9e+99  distance to in2 =9e+99
x=0  y=0  z=0 is outside  distance to in1 =10  distance to in2 =7.07107
x=0  y=0  z=5 is outside  distance to in1 =10  distance to in2 =7.07107
x=0  y=0  z=10 is outside  distance to in1 =10  distance to in2 =7.07107
x=0  y=0  z=15 is outside  distance to in1 =35  distance to in2 =24.7487
x=0  y=0  z=20 is outside  distance to in1 =60  distance to in2 =42.4264
x=0  y=0  z=25 is outside  distance to in1 =45  distance to in2 =31.8198
x=0  y=0  z=30 is outside  distance to in1 =30  distance to in2 =21.2132
x=0  y=0  z=35 is outside  distance to in1 =30  distance to in2 =21.2132
x=0  y=0  z=40 is outside  distance to in1 =30  distance to in2 =21.2132
x=0  y=0  z=45 is outside  distance to in1 =9e+99  distance to in2 =9e+99


==================================================
Pdep is (-110, -110, z)
Dir is (1, 1, 0)
x=-110  y=-110  z=-5 is outside  distance to in =9e+99
x=-110  y=-110  z=0 is outside  distance to in =120.208
x=-110  y=-110  z=5 is outside  distance to in =120.208
x=-110  y=-110  z=10 is outside  distance to in =120.208
x=-110  y=-110  z=15 is outside  distance to in =102.53
x=-110  y=-110  z=20 is outside  distance to in =84.8528
x=-110  y=-110  z=25 is outside  distance to in =84.8528
x=-110  y=-110  z=30 is outside  distance to in =84.8528
x=-110  y=-110  z=35 is outside  distance to in =91.9239
x=-110  y=-110  z=40 is outside  distance to in =98.9949
x=-110  y=-110  z=45 is outside  distance to in =9e+99


